### PR TITLE
Enable fp8 inference for Llava-Next and add Fused_SDPA

### PR DESCRIPTION
**Pull request by tthakkal**
# What does this PR do?

- Add fp8 support for llava_next models llava-v1.6-mistral-7b-hf and llava-v1.6-vicuna-13b-hf.
- Add FusedSDPA support for Llava-1.5-7b, Llava-1.5-13b models

Here is an example to measure the tensor quantization statistics on llava-v1.6-mistral-7b-hf:
```bash
QUANT_CONFIG=./quantization_config/maxabs_measure.json python run_pipeline.py \
--model_name_or_path llava-hf/llava-v1.6-mistral-7b-hf \
--image_path "https://llava-vl.github.io/static/images/view.jpg" \
--use_hpu_graphs \
--bf16
```
07/02/2024 18:20:09 - INFO - __main__ - result = [[{'generated_text': '[INST]  \nWhat is shown in this image? [/INST] The image shows a wooden dock extending into a calm body of water, likely a lake. In the background, there are mountains with snow-capped peaks, suggesting a high-altitude location. The sky is partly cloudy, and the overall scene is serene and picturesque, possibly indicating a location that is popular for outdoor activities such as hiking, fishing, or boating. '}]], time = 2605.2694872021675ms, Throughput (including tokenization) = 29.939321203874847 tokens/second

Here is an example to quantize the model based on previous measurements for llava-v1.6-mistral-7b-hf:
```bash
QUANT_CONFIG=./quantization_config/maxabs_quant.json python run_pipeline.py \
--model_name_or_path llava-hf/llava-v1.6-mistral-7b-hf \
--image_path "https://llava-vl.github.io/static/images/view.jpg" \
--use_hpu_graphs \
--bf16
```
07/02/2024 18:21:45 - INFO - __main__ - result = [[{'generated_text': '[INST]  \nWhat is shown in this image? [/INST] The image shows a wooden dock extending into a calm body of water, likely a lake. The dock is surrounded by a forested area with coniferous trees, and in the background, there are mountains with snow-capped peaks, suggesting a high-altitude location. The sky is partly cloudy, and the overall scene is serene and natural. '}]], time = 742.5564736127853ms, Throughput (including tokenization) = 96.96232213786507 tokens/second

Here is an example to measure the tensor quantization statistics on llava-v1.6-vicuna-13b-hf:
```bash
QUANT_CONFIG=./quantization_config/maxabs_measure.json python run_pipeline.py \
--model_name_or_path llava-hf/llava-v1.6-vicuna-13b-hf \
--image_path "https://llava-vl.github.io/static/images/view.jpg" \
--use_hpu_graphs \
--bf16
```
07/02/2024 18:26:26 - INFO - __main__ - result = [[{'generated_text': "A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions. USER:  \nWhat is shown in this image? ASSISTANT: The image shows a serene lakeside scene. There is a wooden dock extending into a calm body of water, which appears to be a lake. The lake is surrounded by a forested area with tall trees, and in the background, there are mountains with snow on their peaks, suggesting a cooler climate or altitude. The sky is partly cloudy, and the overall atmosphere is peaceful and natural."}]], time = 8267.195312678814ms, Throughput (including tokenization) = 10.03968055196528 tokens/second

Here is an example to quantize the model based on previous measurements for llava-v1.6-vicuna-13b-hf:
```bash
QUANT_CONFIG=./quantization_config/maxabs_quant.json python run_pipeline.py \
--model_name_or_path llava-hf/llava-v1.6-vicuna-13b-hf \
--image_path "https://llava-vl.github.io/static/images/view.jpg" \
--use_hpu_graphs \
--bf16
```
07/02/2024 18:28:30 - INFO - __main__ - result = [[{'generated_text': "A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions. USER:  \nWhat is shown in this image? ASSISTANT: The image shows a serene and picturesque scene of a lake with a wooden dock or pier extending into the water. The lake is surrounded by a forested area with tall trees, and in the background, there are mountains with snow-capped peaks, suggesting it might be a colder season or the mountains are high enough to retain snow. The sky is overcast with clouds, giving the scene a calm and tranquil atmosphere."}]], time = 1261.9115926325321ms, Throughput (including tokenization) = 69.73547157643519 tokens/second

### Inference with FusedSDPA

More information on Using Fused Scaled Dot Product Attention (FusedSDPA) here:
https://docs.habana.ai/en/latest/PyTorch/Model_Optimization_PyTorch/Optimization_in_PyTorch_Models.html?highlight=fusedsdpa#using-fused-scaled-dot-product-attention-fusedsdpa

Currently FusedSDPA only works for Llava-1.5-7b, Llava-1.5-13b models

Use the following commands to run Llava-1.5-7b inference with FusedSDPA

BF16:
```bash
python3 run_pipeline.py \
    --model_name_or_path llava-hf/llava-1.5-7b-hf \
    --use_hpu_graphs \
    --bf16 \
    --use_flash_attention
```

FP8:
Here is an example to measure the tensor quantization statistics on Llava-1.5-7b:
```bash
QUANT_CONFIG=./quantization_config/maxabs_measure.json python run_pipeline.py \
--model_name_or_path llava-hf/llava-1.5-7b-hf \
--image_path "https://llava-vl.github.io/static/images/view.jpg" \
--use_hpu_graphs \
--bf16 \
--use_flash_attention
```

Here is an example to quantize the model based on previous measurements for Llava-1.5-7b:
```bash
QUANT_CONFIG=./quantization_config/maxabs_quant.json python run_pipeline.py \
--model_name_or_path llava-hf/llava-1.5-7b-hf \
--image_path "https://llava-vl.github.io/static/images/view.jpg" \
--use_hpu_graphs \
--bf16 \
--use_flash_attention
```

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: examples/image-to-text/run_pipeline.py [hunk 1]
```diff
@@ -109,7 +115,7 @@ def main():
         args.prompt = "<image>\nUSER: What's the content of the image?\nASSISTANT:"
     elif args.prompt is None and model_type == "llava_next":
         args.prompt = "[INST] <image>\nWhat is shown in this image? [/INST]"
-        if args.model_name_or_path == "llava-hf/llava-v1.6-vicuna-13b-hf":
+        if "vicuna" in args.model_name_or_path:
```
Comment (mandy-li) [inline examples/image-to-text/run_pipeline.py:118]: why this change?

File: examples/image-to-text/README.md [hunk 2]
```diff
@@ -34,24 +34,24 @@ Models that have been validated:
...
 ### Running with FP8
 
-Llava-1.5-7b and Llava-1.5-13b in FP8 are enabled using the Quantization Toolkit (HQT), which provides model measurement and quantization capabilities in PyTorch.
+Llava-1.5-7b, Llava-1.5-13b, llava-v1.6-mistral-7b-hf and llava-v1.6-vicuna-13b-hf in FP8 are enabled using the Quantization Toolkit (HQT), which provides model measurement and quantization capabilities in PyTorch.
 
 More information on enabling fp8 in SynapseAI is available here:
 https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_FP8.html
 
-Here is an example to measure the tensor quantization statistics on Llava-1.5-7b:
+Here is an example to measure the tensor quantization statistics on llava-v1.6-mistral-7b-hf:
```
Comment (mandy-li) [inline examples/image-to-text/README.md:42]: why remove the examples for llava-1.5? pls just add your examples for llava-1.6

**Review by mandy-li — CHANGES_REQUESTED** [review]

File: examples/image-to-text/run_pipeline.py [hunk 3]
```diff
@@ -91,6 +91,12 @@ def main():
         action="store_true",
         help="Whether to ignore eos, set False to disable it.",
     )
+    parser.add_argument(
+        "--use_flash_attention",
+        action="store_true",
+        help="Whether to enable Habana Flash Attention, provided that the model supports it.",
+    )
+
```
Comment (schoi-habana) [inline examples/image-to-text/run_pipeline.py:99]: need --causal_attention_mask

File: examples/image-to-text/run_pipeline.py [hunk 4]
```diff
@@ -149,6 +155,7 @@ def main():
         "hpu_graphs": args.use_hpu_graphs,
         "max_new_tokens": args.max_new_tokens,
         "ignore_eos": args.ignore_eos,
+        "use_flash_attention": args.use_flash_attention,
```
Comment (schoi-habana) [inline examples/image-to-text/run_pipeline.py:158]: need to pass args.causal_attention_mask, too

File: examples/image-to-text/README.md [hunk 5]
```diff
@@ -34,24 +34,24 @@ Models that have been validated:
...
 
-Here is an example to quantize the model based on previous measurements for Llava-1.5-7b:
+Here is an example to quantize the model based on previous measurements for llava-v1.6-mistral-7b-hf:
 ```bash
 QUANT_CONFIG=./quantization_config/maxabs_quant.json python run_pipeline.py \
---model_name_or_path llava-hf/llava-1.5-7b-hf \
+--model_name_or_path llava-hf/llava-v1.6-mistral-7b-hf \
 --image_path "https://llava-vl.github.io/static/images/view.jpg" \
 --use_hpu_graphs \
 --bf16
```
Comment (schoi-habana) [inline examples/image-to-text/README.md:57]: seems the main change is to add fused sdpa but none of the examples use it. any reason not to add an example with flash attention enabled?

**Review by schoi-habana — CHANGES_REQUESTED** [review]

File: examples/image-to-text/README.md (hunk 2 above)
Comment (tthakkal) [reply examples/image-to-text/README.md:42]: Will add

File: examples/image-to-text/README.md (hunk 5 above)
Comment (tthakkal) [reply examples/image-to-text/README.md:57]: Because of accuracy issue, its not producing right result with sdpa, going to create ticket on that

File: examples/image-to-text/run_pipeline.py (hunk 1 above)
Comment (tthakkal) [reply examples/image-to-text/run_pipeline.py:118]: To enable llava-hf/llava-v1.6-vicuna-7b-hf model

File: examples/image-to-text/run_pipeline.py (hunk 4 above)
Comment (tthakkal) [reply examples/image-to-text/run_pipeline.py:158]: Vision model already have parameter for same

File: optimum/habana/transformers/models/clip/modeling_clip.py [hunk 6]
```diff
@@ -16,3 +34,306 @@ def forward(self, pixel_values: torch.FloatTensor) -> torch.Tensor:
...
+
+class Softmax(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x, dim=None, invAttnHead=None):
+        return torch.ops.hpu.softmax_fp8(x, dim, None, None, invAttnHead)
+
+
+# Below code is copied from# https://github.com/huggingface/transformers/blob/v4.41.2/src/transformers/models/clip/modeling_clip.py
```
Comment (mandy-li) [inline optimum/habana/transformers/models/clip/modeling_clip.py:64]: please add python doc to each class/function to explain what is original class/function, what is different from original, i.e what changes you made by creating this new class/function

**Review by mandy-li — CHANGES_REQUESTED** [review]

File: examples/image-to-text/README.md (hunk 2 above)
Comment (mandy-li) [reply examples/image-to-text/README.md:42]: Change llava to Llava in your text

File: examples/image-to-text/README.md [hunk 7]
```diff
@@ -34,24 +34,24 @@ Models that have been validated:
...
+Llava-1.5-7b, Llava-1.5-13b, llava-v1.6-mistral-7b-hf and llava-v1.6-vicuna-13b-hf in FP8 are enabled using the Quantization Toolkit (HQT), which provides model measurement and quantization capabilities in PyTorch.
 
 More information on enabling fp8 in SynapseAI is available here:
 https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_FP8.html
 
-Here is an example to measure the tensor quantization statistics on Llava-1.5-7b:
+Here is an example to measure the tensor quantization statistics on llava-v1.6-mistral-7b-hf:
 ```bash
 QUANT_CONFIG=./quantization_config/maxabs_measure.json python run_pipeline.py \
---model_name_or_path llava-hf/llava-1.5-7b-hf \
```
Comment (mandy-li) [inline examples/image-to-text/README.md:45]: Do not remove the example for llava-1.5, add llava-next examples after llava-1.5

File: examples/image-to-text/README.md [hunk 8]
```diff
@@ -34,24 +34,24 @@ Models that have been validated:
...
 --image_path "https://llava-vl.github.io/static/images/view.jpg" \
 --use_hpu_graphs \
 --bf16
 ```
 
-Here is an example to quantize the model based on previous measurements for Llava-1.5-7b:
+Here is an example to quantize the model based on previous measurements for llava-v1.6-mistral-7b-hf:
 ```bash
 QUANT_CONFIG=./quantization_config/maxabs_quant.json python run_pipeline.py \
---model_name_or_path llava-hf/llava-1.5-7b-hf \
```
Comment (mandy-li) [inline examples/image-to-text/README.md:54]: do not remove llava-1.5 examples

**Review by mandy-li — CHANGES_REQUESTED** [review]
pls address the comments

File: optimum/habana/transformers/models/clip/modeling_clip.py [hunk 9]
```diff
@@ -16,3 +34,306 @@ def forward(self, pixel_values: torch.FloatTensor) -> torch.Tensor:
...
+        key_states = key_states.view(*proj_shape)
+        value_states = value_states.view(*proj_shape)
+
+        src_len = key_states.size(1)
+        if FusedSDPA and use_flash_attention:
+            import habana_frameworks.torch.hpu as ht
+
+            if tgt_len == 1:
+                # next token
+                use_recompute = True if os.getenv("QUANT_CONFIG", "") else False
```
[the remark this answers is not in the document]
Comment (tthakkal) [reply optimum/habana/transformers/models/clip/modeling_clip.py]: Changed and simplified logic

File: examples/image-to-text/README.md [hunk 10]
```diff
@@ -125,11 +125,50 @@ QUANT_CONFIG=./quantization_config/maxabs_measure.json python run_pipeline.py \
 --bf16
 ```
 
-Here is an example to quantize the model based on previous measurements for Llava-v1.6-mistral-7b-hf:
+Here is an example to quantize the model based on previous measurements for Llava-v1.6-mistral-7b:
```
Comment (mandy-li) [inline examples/image-to-text/README.md:128]: Llava-v1.6-vicuna-13b

File: examples/image-to-text/README.md [hunk 11]
```diff
@@ -125,11 +125,50 @@ QUANT_CONFIG=./quantization_config/maxabs_measure.json python run_pipeline.py \
...
 QUANT_CONFIG=./quantization_config/maxabs_quant.json python run_pipeline.py \
 --model_name_or_path llava-hf/llava-v1.6-vicuna-13b-hf \
 --image_path "https://llava-vl.github.io/static/images/view.jpg" \
 --use_hpu_graphs \
 --bf16
 ```
+
+### Inference with FusedSDPA
+
+More information on Using Fused Scaled Dot Product Attention (FusedSDPA) here:
```
Comment (mandy-li) [inline examples/image-to-text/README.md:139]: Habana FusedSDPA is a fused and optimized implementation of torch.nn.functional.scaled_dot_product_attention() for Gaudi. For more details, refer to [Habana online document](https://docs.habana.ai/en/latest/PyTorch/Model_Optimization_PyTorch/Optimization_in_PyTorch_Models.html?highlight=fusedsdpa#using-fused-scaled-dot-product-attention-fusedsdpa).

**Review by mandy-li — CHANGES_REQUESTED** [review]
Looks much better now. Some minor comments to address.

**Review by mandy-li — CHANGES_REQUESTED** [review]
Looks much better. Minor changes are suggested

File: optimum/habana/transformers/models/clip/modeling_clip.py [hunk 12]
```diff
@@ -16,3 +34,310 @@ def forward(self, pixel_values: torch.FloatTensor) -> torch.Tensor:
...
+
+            if tgt_len == 1 or causal_attention_mask is None:
+                with ht.sdp_kernel(enable_recompute=True):
+                    attn_output = self.fused_scaled_dot_product_attention(
+                        query_states, key_states, value_states, attention_mask, self.dropout, False, 1, "fast"
+                    )
+            elif causal_attention_mask is not None:
+                with ht.sdp_kernel(enable_recompute=True):
+                    attn_output = self.fused_scaled_dot_product_attention(
+                        query_states, key_states, value_states, causal_attention_mask, self.dropout, True, 1, "fast"
```
[the remark this answers is not in the document]
Comment (tthakkal) [reply optimum/habana/transformers/models/clip/modeling_clip.py:110]: Difference is setting is_causal to False in fused_scaled_dot_product_attention call

File: examples/image-to-text/README.md [hunk 13]
```diff
@@ -56,3 +96,65 @@ QUANT_CONFIG=./quantization_config/maxabs_quant.json python run_pipeline.py \
...
+QUANT_CONFIG=./quantization_config/maxabs_quant.json python run_pipeline.py \
+--model_name_or_path llava-hf/llava-v1.6-vicuna-13b-hf \
+--image_path "https://llava-vl.github.io/static/images/view.jpg" \
+--use_hpu_graphs \
+--bf16
+```
+
+### Inference with FusedSDPA
+
+Habana FusedSDPA is a fused and optimized implementation of torch.nn.functional.scaled_dot_product_attention() for Gaudi. For more details, refer to [Habana online document](https://docs.habana.ai/en/latest/PyTorch/Model_Optimization_PyTorch/Optimization_in_PyTorch_Models.html?highlight=fusedsdpa#using-fused-scaled-dot-product-attention-fusedsdpa). Currently FusedSDPA only works with BF16 precision.
```
[the remark this answers is not in the document]
Comment (tthakkal) [reply examples/image-to-text/README.md:139]: updated

File: optimum/habana/transformers/models/clip/modeling_clip.py [hunk 14]
```diff
@@ -16,3 +34,310 @@ def forward(self, pixel_values: torch.FloatTensor) -> torch.Tensor:
...
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        causal_attention_mask: Optional[torch.Tensor] = None,
+        output_attentions: Optional[bool] = False,
+        use_flash_attention: Optional[bool] = False,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        """
+        Copied from CLIPAttention.forward: https://github.com/huggingface/transformers/blob/ab0f050b42d903f34d6eb97f3f8c0c07f0517ad2/src/transformers/models/clip/modeling_clip.py
+        The only differences are:
+        - add new args use_flash_attention
```
[the remark this answers is not in the document]
Comment (tthakkal) [reply optimum/habana/transformers/models/clip/modeling_clip.py:84]: both are same modified comment to use_flash_attention to enable FusedSDPA

File: optimum/habana/transformers/models/clip/modeling_clip.py (hunk 12 above)
Comment (schoi-habana) [inline optimum/habana/transformers/models/clip/modeling_clip.py:110]: are you sure about using causal_attention_mask as an input to the fused sdpa?
from other examples with is_causal=true, None is passed instead.

File: optimum/habana/transformers/models/clip/modeling_clip.py [hunk 15]
```diff
@@ -16,3 +34,310 @@ def forward(self, pixel_values: torch.FloatTensor) -> torch.Tensor:
...
+        if FusedSDPA and use_flash_attention:
+            import habana_frameworks.torch.hpu as ht
+
+            if tgt_len == 1 or causal_attention_mask is None:
+                with ht.sdp_kernel(enable_recompute=True):
+                    attn_output = self.fused_scaled_dot_product_attention(
+                        query_states, key_states, value_states, attention_mask, self.dropout, False, 1, "fast"
+                    )
+            elif causal_attention_mask is not None:
+                with ht.sdp_kernel(enable_recompute=True):
```
Comment (schoi-habana) [inline optimum/habana/transformers/models/clip/modeling_clip.py:108]: enabling recompute helps memory consumption, but may harm performance. i guess that's why it's a switch for other models. do we still want it to be true always? @mandy-li

Comment (mandy-li) [reply optimum/habana/transformers/models/clip/modeling_clip.py:108]: @schoi-habana ,why harm perf? it just doesn't cache activation and recompute wont happen anyway because this is inference (ie. no bwd pass), right?

**Review by mandy-li — APPROVED** [review]
LGTM

**Review by schoi-habana — APPROVED** [review]

**Comment by mandy-li** [comment]
@regisss , please review and merge if it is ok for you. thanks

**Review by regisss — APPROVED** [review]
